src/riak_pb_codec.erl

```diff
@@ -73,21 +77,66 @@
 %% Bucket properties that are commit hooks have this format.
 -type commit_hook_property() :: [ {struct, [{commit_hook_field(), binary()}]} ].
 
+%% The protobuf message ID for a timeseries query response, needed to
+%% allow our nif interface to identify the message for which we wish
+%% to perform optimized decoding.
+-define(TIMESERIES_QUERY_RESP, 91).
+
+-spec init() -> any().
+init() ->
+    SoName = case code:priv_dir(?MODULE) of
+                 {error, bad_name} ->
+                     case code:which(?MODULE) of
+                         Filename when is_list(Filename) ->
+                             filename:join([filename:dirname(Filename),"../priv", "riak_pb_codec"]);
+                         _ ->
+                             filename:join("../priv", "riak_pb_codec")
+                     end;
+                 Dir ->
+                     filename:join(Dir, "riak_pb_codec")
+             end,
+    erlang:load_nif(SoName, 0),
+
+    ok.
+
+encode_tsputreq(Msg) ->
```

Comment:
This is NIF magic.  The function name encode_tsputreq is listed within the NIF as a function to replace.  When the NIF exists, lines 103 - 105 never get executed.  Execution passes to the NIF.  Should the NIF not exists (open source version) or fail to load, then 103 to 105 are the "backup code".  Yes, the code is intentionally a direct copy from encode/1 since that is the correct action if the NIF does not perform the work.  Good question!